torch_xla/csrc/xla_sharding_util.cpp

```diff
@@ -413,9 +433,9 @@ ShardingUtil::GetShardIndicesForDevices(
           continue;
         }
         int64_t n_j = offset % tile_shape[j];
-        int start = n_j * shard_shape[j];
-        // Clamp the end of the slice to the tensor shape to accurately reflect
+        // Clamp the slice bounds to the tensor shape to accurately reflect
         // the shard size without padding.
+        int start = std::min(n_j * shard_shape[j], tensor_shape[j]);
```

Comment:
I would call this a latent bug, but it wasn't breaking anything because torch indexing handles negative-length indices as though they were empty. It just breaks the expectation that `stop - start` reflects the size of the unpadded shard, which we rely on in distributed checkpointing.

You're right - these index slices will end up empty, but this is the desired outcome when the shard consists entirely of padding.